### ISSUE TITLE
Injecting user.data through Fn.File always triggers deploy

### ISSUE THREAD

**Issue by jin-ahn** [hashicorp/terraform-cdk#1996]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions
cdktf & Language Versions

cdktf - 0.12.0
node - 18.6.0
typescript

### Affected Resource(s)

I launch a nodegroup using the customtemplate option to add user.data through a file. However, when doing so terraform always recognizes there is a change when nothing is changed. I think it's because we're basing in a base64 while each time so it assumes there is something different.
```
      const userDataFile = config.customTemplate.userData ?? "./userData.tpl";
      const launchTemplate = new ec2.LaunchTemplate(
        this,
        this.nodeGroupName + "-template",
        {
          name: this.nodeGroupName + "-template",
          updateDefaultVersion: true,
          lifecycle: {
            createBeforeDestroy: true,
          },
          tags: this.ngtag,
          userData: Fn.filebase64(userDataFile),
          blockDeviceMappings: [{
            deviceName: "/dev/xvda",
            ebs: {
              volumeSize: config.diskSize ?? 20,
            },
          }],
        },
      );
      const ltName = launchTemplate.nameInput;
      const defaultNGConifg = {
        clusterName: this.clusterName,
        instanceTypes: this.instanceTypes,
        nodeRoleArn: this.nodeGroupRoleArn,
        nodeGroupNamePrefix: this.nodeGroupName,
        subnetIds: this.subnetIds,
        launchTemplate: {
          name: ltName,
          version: "$Default",
        },
        amiType: this.amiType ?? "AL2_x86_64",
        scalingConfig: {
          desiredSize: config.desiredSize ?? 1,
          maxSize: config.maxSize ?? 2,
          minSize: config.minSize ?? 1,
        },
        lifecycle: {
          createBeforeDestroy: true,
          ignoreChanges: [
            "scaling_config[0].desired_size",
          ],
        },
        dependsOn: [launchTemplate],
        labels: config.labels,
        tags: this.ngtag,
      };
      this.nodeGroup = new eks.EksNodeGroup(
        this,
        this.nodeGroupName,
        defaultNGConifg,
```
### Expected Behavior

Running cdktf deploy should show that nothing changed

### Actual Behavior

Running cdktf deploy says there a change happened and needs a deploy even though nothing changed 

### Steps to Reproduce

Use fn.File to inject into a custom template for nodegroups. Run consecutive cdktf deploys.

**Comment by DanielMSchmidt** [comment]
Do I understand it correctly that the file itself changes? If so then something did change and a deploy needs to occur.

**Comment by jin-ahn** [comment]
> Do I understand it correctly that the file itself changes? If so then something did change and a deploy needs to occur.

No the issue is that nothing changed at all but it says a deploy needs to occur.

**Comment by ansgarm** [comment]
I think the issue might be within the AWS Terraform provider. This issue seems related: https://github.com/hashicorp/terraform-provider-aws/issues/5011.
Could be something like [this](https://github.com/hashicorp/terraform-provider-aws/issues/5011#issuecomment-878542063) for example.

**Comment by jin-ahn** [comment]
@ansgarm thanks for the response. The suggested fix in the posts you provided is to switch to using `Fn.templateFile` instead of Fn.File. 

Trying to implement i came across a new issue:
`Fn.templateFile(path, var)` requires you to pass in a string that points the file you're trying to use. However, in cdktf I am not able to use ${path.cwd} as stated in this[ ticket](https://github.com/hashicorp/terraform-cdk/issues/1997). The suggested fix for ${path.cwd} is to use Assets as stated [HERE ](https://github.com/hashicorp/terraform-cdk/issues/1997#issuecomment-1205383291).

The problem is using Assets means cdktf creates a new asset file in during the synth. But `Fn.templateFile` only allows files that exist at the begining of the run not files generated by run as stated [HERE](https://www.terraform.io/language/functions/templatefile#lists)

In summary, if my understanding is correct, because cdktf does not properly support having "${path.cwd}", I will never be able to use Fn.templateFile ever.

Any thoughts?

**Comment by jin-ahn** [comment]
The suggested fix is to switch to using `Fn.templateFile`. Below is my attempt to implement it 

```
      const userFile = path.resolve(__dirname, config.customTemplate.userData ?? "./userData.tftpl")
      const launchTemplate = new ec2.LaunchTemplate(
        this,
        this.nodeGroupName + "-template",
        {
          name: this.nodeGroupName + "-template",
          updateDefaultVersion: true,
          lifecycle: {
            createBeforeDestroy: true,
          },
          tags: this.ngtag,
          userData: Fn.filebase64(Fn.templatefile(userFile, config.customTemplate.templateConfig)),
          blockDeviceMappings: [{
            deviceName: "/dev/xvda",
            ebs: {
              volumeSize: config.diskSize ?? 20,
              // encrypted: config.diskEncryption ?? "false",
              // kmsKeyId: config.diskEncryptionKey ?? ""
            },
          }],
        },
      );
```

When I run it I get the following error 
```
Invalid value for "path" parameter: no file exists at "#!/bin/sh\n\naws s3
           │ --region........REDACTED
           ....
           ; this function
           │ works only with files that are distributed as part of the configuration
           │ source code, so if this file will be created by a resource in this
           │ configuration you must instead obtain this result from an attribute of that
           │ resource.
```

the error message actually spits out the contents of the file I'm trying to feed it so I know it "found" the file, but it's complaining that it doesn't exist.

any ideas?

**Comment by ansgarm** [comment]
Hi @jin-ahn I got an notification email with some comment of yours about assets. Did you delete that comment?
Because you should be able to use assets here as well. As far as Terraform is concerned those files exist before Terraform is invoked as the CDKTF CLI copies the assets before invoking the Terraform CLI.

**Comment by ansgarm** [comment]
(Well the synth step copies them even)

**Comment by jin-ahn** [comment]
@ansgarm  I did have a write-up about assets, but I realized I might have been overcomplicating with using assets when I could just rely on `path.resolve`. However, in both cases, I'm facing the issue where I get the error where Fn.templateFile says the file does not exist. So im still having issue from my above comment

**Comment by DanielMSchmidt** [comment]
[templatefile](https://www.terraform.io/language/functions/templatefile) takes a path to a file, you are passing the content directly, I think that is why it fails here

**Comment by jin-ahn** [comment]
> [templatefile](https://www.terraform.io/language/functions/templatefile) takes a path to a file, you are passing the content directly, I think that is why it fails here
that is not true. I am passing in the path, the error prints out the path so i know its getting put in correctly.  but it still says its not found

**Comment by ansgarm** [comment]
Ah! The error seems to origin in the `filebase64` Terraform function which
> filebase64 reads the contents of a file at the given path and returns them as a base64-encoded string.

whereas `templatefile` already returns the contents of a file. Switching to `Fn.base64encode` should do the trick then!

**Comment by jin-ahn** [comment]
that was fixed to get it working! thank you @ansgarm 

However the original issue of cdktf constantly deploying still exists, but discovered it's related to launch_templates not Fn.File so closing this ticket to start a new thread
